### PR TITLE
[WIP] Babel plugin for preprocessing

### PR DESCRIPTION
**Pull request by kof**
Issue https://github.com/cssinjs/jss/issues/579

### CODE REVIEW

File: packages/babel-plugin/src/index.js [hunk 1]
```diff
@@ -0,0 +1,122 @@
+import {declare} from '@babel/helper-plugin-utils'
+import {create as createJss} from 'jss'
+
+const rawRuleName = '@raw'
+const defaultIdentifiers = ['createStyleSheet', 'injectSheet']
+
+export default declare(({types: t, ...api}, {identifiers = defaultIdentifiers, jssOptions}) => {
+  api.assertVersion(7)
```
Comment (HenriBeck) [inline packages/babel-plugin/src/index.js]: Should we maybe update our own babel packages to v7 as well? Would make testing easier.

Comment (kof) [reply packages/babel-plugin/src/index.js]: sure, if you get time, v6 and 7 are largely compatible

**Comment by kof** [comment]
@cssinjs/core anyone wants to review this one? I am in the testing phase yet, but I want already start with making this code high quality readable

**Comment by HenriBeck** [comment]
Can you rename the directory from `babel-plugin` to `babel-plugin-jss`?

File: packages/babel-plugin/src/testUtils.js [hunk 2]
```diff
@@ -0,0 +1,14 @@
+import * as babel7 from '@babel/core'
```
Comment (HenriBeck) [inline packages/babel-plugin/src/testUtils.js]: Maybe rename to just babel?

Comment (kof) [reply packages/babel-plugin/src/testUtils.js]: I am working on tests to ensure babel6 works, so it will be needed

**Comment by kof** [comment]
> Can you rename the directory from babel-plugin to babel-plugin-jss?

yep

**Comment by kof** [comment]
> Also, can we have one standard for where the test files should be located? Because right now sometimes *.test.js is used and sometimes all of the tests are inside a tests folder.

There are 2 different testing scenarios. One is unit tests, where I actually test a specific module. The other one is more like an integration tests. In this case I 100% did integration tests and I had them all in index.test.js but then the amount of tests grew and file became unreadable, so I split them up in separate suits.

I think its ok to not have this consistency. If you are testing a specific module, then it should be module.test.js, otherwise if a test uses multiple modules it should go into __tests__. Both patterns are supported by jest by default without changing the config.

**Comment by kof** [comment]
Remaining todo:
-  find a way to identify function calls without relying on a name, because this doesn't work after certain babel plugins, it be `something.default()`
- find a way to extract imported static variables
- evaluate module code in a separate context instead of `eval`
- add tests ensuring the proper variable scoping when evaluating (use same variable name to detect wrong scoping)
- use evaluation more instead of nodes traversing, seems like a more robust appraoch

File: packages/babel-plugin-jss/package.json [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+    "@babel/core": "^7.0.0-beta.54",
+    "babel-core": "^6.26.3",
+    "babel-plugin-syntax-decorators": "^6.13.0",
+    "jest": "^23.4.0"
+  },
+  "dependencies": {
+    "@babel/helper-plugin-utils": "^7.0.0-beta.53",
+    "babel-generator": "^6.26.1",
+    "babel-types": "^6.26.0",
+    "jss": "^9.8.7",
```
Comment (HenriBeck) [inline packages/babel-plugin-jss/package.json]: Shouldn't jss be a peer dependency?

Comment (kof) [reply packages/babel-plugin-jss/package.json]: seems valid

**Comment by jahglow** [comment]
guys, any progress  on this? Like css extraction I mean...

**Pull request closed without merging**